### ISSUE TITLE
LAN interface

### ISSUE THREAD

**Issue by Aehmlo** [Aehmlo/lifxi#7]
We currently support the HTTP API, which is awesome, but let's also add support for the LAN version.

**Comment by eminence** [comment]
Hello!  Author of [`lifx-core`](https://crates.io/crates/lifx-core) here.   Happy to lend help and/or code to this.

**Comment by Aehmlo** [comment]
@eminence Awesome! I'm going to work on finishing [this milestone](https://github.com/Aehmlo/lifxi/milestone/1) first, and then I'll be ready to transition to LAN work. I'll take a look at `lifx-core` and see how it might fit into this crate.

**Comment by Aehmlo** [comment]
@eminence I've been busy working on [other stuff](https://github.com/Aehmlo/adm), but I think it's about time for me to get back to looking at the LAN interface. I think that `lifx-core` provides a pretty good starting point for LAN work, and it might make the most sense for me to contribute to that where necessary and pull that crate into this crate as a dependency (implementing the actual over-the-wire stuff in this crate, probably). Does this sound like a sane course of action to you, or is there another route you'd prefer/like me to consider?

**Comment by eminence** [comment]
Yeah, that sounds like good plan.  Since `lifx-core` doesn't handle any networking stuff, there's a need for a higher-level crate to manage that (like `lifxi`).  The primary thing a crate like this needs to do is:

* Manage bulb state
* Manage the network communication

Bulb state is important for things like bulb discovery and waiting for message acknowledgements.  For example, to do bulb discovery, you have to:

1. Bind to UDP port 56700
2. Send a broadcast package with the `GetService` message
3. Wait for replies, to learn about the bulb's ID and IP address.
4. Optionally send follow-up messages to get bulb name/group information (and wait for replies).

Another avenue for development is a local LIFX web server for control local bulbs on the LAN.  This is useful for people who want multiple apps to control the bulbs.  Since multiple programs can't both bind to port 56700 at the same time, only 1 program will be able to receive replies at a time.  Thus a single daemon service with a web interface to control the bulbs could be useful.

**Comment by georgefst** [comment]
Any progress on this? I'd be willing to lend a hand in any way I can if it's just down to a lack of time.

**Comment by Aehmlo** [comment]
Hey @georgefst! I don’t have any news here: I was planning to revisit this crate once `async/await` stabilized, but haven’t gotten around to it just yet. I’d be happy to have the help, if you want to send a PR or whatever!

**Comment by georgefst** [comment]
Hey there! Having looked at `lifx-core` it's actually fine for my, fairly limited, use case.

Unless I find myself with a lot of free time, implementing a higher-level interface is probably best left to someone who knows a bit more about networking (and Rust...).
